src/main/scala/org/zalando/spark/jsonschema/SchemaConverter.scala

```diff
@@ -69,43 +79,45 @@ object SchemaConverter {
-    val source = scala.io.Source.fromFile(filePath)
-    try Json.parse(source.getLines.mkString)
-    finally source.close()
+    val source = Source.fromFile(filePath)
+    try {
+      Json.parse(source.getLines.mkString)
+    } finally {
+      source.close()
+    }
   }
 
   @tailrec
-  def convertJsonStruct(schema: StructType, json: JsValue, jsonKeys: Set[String]): StructType = {
-    if (jsonKeys.nonEmpty) {
-      val enrichedSchema = addJsonField(schema, (json \ jsonKeys.head).as[JsValue])
-      convertJsonStruct(enrichedSchema, json, jsonKeys.drop(1))
-    } else { schema }
+  private def convertJsonStruct(schema: StructType, json: JsObject, jsonKeys: List[String]): StructType = {
+    jsonKeys match {
+      case Nil => schema
+      case head :: tail =>
+        val enrichedSchema = addJsonField(schema, (json \ head).as[JsValue])
+        convertJsonStruct(enrichedSchema, json, tail)
+    }
   }
 
-  def addJsonField(schema: StructType, json: JsValue): StructType = {
+  private def addJsonField(schema: StructType, json: JsValue): StructType = {
     val fieldType = getJsonType(json)
-    typeMap(fieldType.typeName) match {
-      case d: DataType => schema.add(getJsonName(json), d, nullable = fieldType.nullable)
+    val (dataType, nullable) = typeMap(fieldType.typeName) match {
+
+      case dataType: DataType =>
+        (dataType, fieldType.nullable)
+
       case ArrayType =>
-        addJsonArrayStructure(schema, json, JsPath \ SchemaArrayContents \ SchemaStructContents)
-      case StructType => addJsonObjectStructure(schema, json, JsPath \ SchemaStructContents)
-    }
-  }
+        val dataType = ArrayType(getDataType(json, JsPath \ SchemaArrayContents \ SchemaStructContents))
+        (dataType, getJsonType(json).nullable)
 
-  private def addJsonStructure(schema: StructType, json: JsValue, contentPath: JsPath)(addStructure: StructType => DataType): StructType = {
-    val content = contentPath.asSingleJson(json).as[JsObject]
-    schema.add(
-      getJsonName(json),
-      addStructure(convertJsonStruct(new StructType, content, content.keys)),
-      nullable = getJsonType(json).nullable
-    )
-  }
+      case StructType =>
+        val dataType = getDataType(json, JsPath \ SchemaStructContents)
+        (dataType, getJsonType(json).nullable)
+    }
 
-  def addJsonObjectStructure(schema: StructType, json: JsValue, contentPath: JsPath): StructType = {
-    addJsonStructure(schema, json, contentPath)(structType => structType)
+    schema.add(getJsonName(json), dataType, nullable = nullable)
   }
 
-  def addJsonArrayStructure(schema: StructType, json: JsValue, contentPath: JsPath): StructType = {
-    addJsonStructure(schema, json, contentPath)(structType => ArrayType(structType))
+  private def getDataType(json: JsValue, contentPath: JsPath): DataType = {
+    val content = contentPath.asSingleJson(json).as[JsObject]
+    convertJsonStruct(new StructType, content, content.keys.toList)
```

Comment:
great! great, looks like you demystified things here a lot.
From my point of view, it does the same thing.
But @hesserp should also check this